### PR TITLE
gdsfactory-gen generator code update

### PR DESCRIPTION
**Pull request by vijayshankarr**
Currently, errors are seen when running both line-res_via-gen and mimcap_gen scripts. 
In this pull request, updates are made in the Copy and CrossSection method calls of Component API, to fix those errors.

Note: There are a few warnings that can still be seen, as follows:
image

### CODE REVIEW

**Comment by msaligane** [comment]
@joamatab FYI.

**Comment by joamatab** [comment]
yes, we are starting to write warning to discourage writing dangerous code with gdsfactory

https://gdsfactory.github.io/gdsfactory/notebooks/common_mistakes.html

All the gdsfactory code in this repo has a lot of room for improvement

Here are some ideas:

- write all your Components into Pcell functions, don't name things manually
- break code into small cell functions
- Don't move components, create references and move the references instead


I would recommend reading the gdsfactory tutorial a few times and try to rewrite the code in a pythonic way

**Comment by msaligane** [comment]
Thanks @joamatab .@vijayshankarr can you please go over this? I was hoping to see some of this in our last Friday meeting.

**Comment by vijayshankarr** [comment]
Sure @msaligane. Last week, I was focussed on getting the code run without any errors. I will get started with rewriting the code with the suggestions mentioned. Thanks.

**Comment by vijayshankarr** [comment]
@msaligane The generator codes have been updated as per the suggestions of @joamatab. The warnings are not seen anymore.

**Comment by joamatab** [comment]
Looks much better 


The cell decorator will also set the name so trying to pass the name in the functions has no effect

**Comment by vijayshankarr** [comment]
Yes, that's right. I have updated the code to remove the component name arguments from all the definitions.

**Comment by msaligane** [comment]
Thanks @vijayshankarr - Is there a way we could add a regression test for this?

**Comment by vijayshankarr** [comment]
@msaligane Is there a regression setup already available, that needs to be used or do we need to create a new one?

**Comment by msaligane** [comment]
@vijayshankarr I would see a gdsok file to compare with as a good regression test and DRC/LVS checks.

@saicharan0112 can you work with Vijay to add this to the CI

**Comment by saicharan0112** [comment]
Of course. Vijay, can we meet tomorrow? I will ping you on slack.

**Comment by saicharan0112** [comment]
@msaligane I met with Vijay last Friday. Basically these generators should be checked for DRC and LVS errors. I came to know from Vijay that currently these gds generation doesn't depend on the PDK. Also not sure how to verify LVS without source netlist. Once I know how to check both, I will add this along with other generators.

@vijayshankarr do you want to add anything here?

**Comment by msaligane** [comment]
@vijayshankarr any update on this?

**Comment by saicharan0112** [comment]
@vijayshankarr @alibillalhammoud please let me know once the drc report generation is done

**Comment by vijayshankarr** [comment]
@saicharan0112 The flow currently generates a DRC report in the result_dir directory

**Comment by vijayshankarr** [comment]
@saicharan0112 
@alibillalhammoud and I just reviewed the dependencies script for merge conflict, please see if we made any errors

**Comment by alibillalhammoud** [comment]
@vijayshankarr and I added the MappedPDK class. We will use MappedPDK to integrate all additional generic generator features. Currently, the generic layers and DRC are included. Mapped PDKs are defined for Sky130 and GF180. There is a constructor for this class which takes a gdsfactory PDK and a generic layer mapping to produce a MappedPDK object.
MappedPDK will appropriately raise and handle errors using pydantic (very similar to how GDSFactory does). It is a child class of the GDSFactory Pdk class.

**Comment by alibillalhammoud** [comment]
@msaligane PR is ready to merge

File: openfasoc/generators/gdsfactory-gen/scripts/mimcap-array/drc-check/magic_commands.tcl
Comment (msaligane) [inline openfasoc/generators/gdsfactory-gen/scripts/mimcap-array/drc-check/magic_commands.tcl:1]: I think it makes more sense to move all these magic cmds and DRC checks in one common folder.

File: openfasoc/generators/gdsfactory-gen/scripts/mimcap-array/drc-check/sky130A/sky130A.magicrc
Comment (msaligane) [inline openfasoc/generators/gdsfactory-gen/scripts/mimcap-array/drc-check/sky130A/sky130A.magicrc:1]: same, this is coming ffrom the PDK and these files are subject to updates. We should find a way to extract them from the PDK each time.

Comment (msaligane) [reply openfasoc/generators/gdsfactory-gen/scripts/mimcap-array/drc-check/sky130A/sky130A.magicrc:1]: @proppy any recommendations?

**Comment by msaligane** [comment]
@alibillalhammoud @vijayshankarr can you please check my comments?

File: openfasoc/generators/gdsfactory-gen/scripts/mimcap-array/drc-check/magic_commands.tcl
Comment (alibillalhammoud) [reply openfasoc/generators/gdsfactory-gen/scripts/mimcap-array/drc-check/magic_commands.tcl:1]: We have not changed any old code. Instead, we moved all existing generators into scripts dir

File: openfasoc/generators/gdsfactory-gen/PDK/gf180_mapped/gf180mcu_drc.lydrc
[the remark this answers is not in the document]
Comment (alibillalhammoud) [reply openfasoc/generators/gdsfactory-gen/PDK/gf180_mapped/gf180mcu_drc.lydrc:1]: (in next PR) We made some edits to the DRC script so it runs full DRC without needing to pass the appropriate flags.

File: openfasoc/generators/gdsfactory-gen/PDK/gf180_mapped/gf180_mapped.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+gf180_mapped_pdk = MappedPDK(
+    name="gf180",
+    glayers=gf180_glayer_mapping,
+    layers=LAYER.dict(),
+    # layer_views = gf180.layers.LAYER_VIEWS
+)
+
+
+# import pathlib
```
[the remark this answers is not in the document]
Comment (vijayshankarr) [reply openfasoc/generators/gdsfactory-gen/PDK/gf180_mapped/gf180_mapped.py]: @msaligane We have removed the comments

File: openfasoc/generators/gdsfactory-gen/scripts/mimcap-array/drc-check/sky130A/sky130A.magicrc
Comment (proppy) [reply openfasoc/generators/gdsfactory-gen/scripts/mimcap-array/drc-check/sky130A/sky130A.magicrc:1]: why not use what's currently in the conda packages?

File: openfasoc/generators/gdsfactory-gen/Makefile
[the remark this answers is not in the document]
Comment (vijayshankarr) [reply openfasoc/generators/gdsfactory-gen/Makefile:1]: @msaligane We have added additional comments on each target.

**Review by msaligane — APPROVED** [review]

**Pull request merged**